nni/common/concrete_trace_utils/counter.py

```diff
@@ -0,0 +1,240 @@
+        rst = super().call_function(target, args, kwargs)
+        return rst, flop_count(target, *args, **kwargs), {}   # FIXME: call_function might also have flops
+
+    def call_module(self, target: Target, args: Tuple[Argument, ...], kwargs: Dict[str, Any]) -> Any:
+        # Execute the method and return the result
+        assert isinstance(target, str)
+        submod = self.fetch_attr(target)
+        return (
+            submod(*args, **kwargs),
+            flop_count(submod, *args, **kwargs),
+            {
+                k: v for k, v in submod.named_parameters()
+            }
+        )
+
+    def get_attr(self, target: Target, args: Tuple[Argument, ...], kwargs: Dict[str, Any]) -> Any:
+        assert isinstance(target, str)
+        rst = self.fetch_attr(target)
+        if isinstance(rst, torch.nn.Parameter):
+            return (
+                rst,
+                0,
+                {target: rst}
+            )
+        return rst, 0, {}
+
+    def summarize(self) -> str:
+        """
+        Summarizes the profiled statistics of the `GraphModule` in
+        tabular format. Note that this API requires the ``tabulate`` module
+        to be installed.
+
+        Returns:
+            str: The summary of the profiled statistics
+        """
+        # https://github.com/pytorch/pytorch/blob/master/torch/fx/graph.py
+        try:
+            from tabulate import tabulate   # pyright: reportUnboundVariable=false
+        except ImportError:
+            print("`summary` relies on the library `tabulate`, "
+                  "which could not be found on this machine. Run `pip "
+                  "install tabulate` to install the library.")
+
+        # Build up a list of summary information for each node
+        node_summaries: List[List[Any]] = []
+
+        for node in self.module.graph.nodes:
+            node: Node
+            n_info = NInfo(node)
+            node_summaries.append([
+                node.op,
+                str(node),
+                _format_memory(n_info.param_size),
+                _format_flops(n_info.flops),    # type: ignore
+            ])
+
+        # Use the ``tabulate`` library to create a well-formatted table
+        # presenting our summary information
+        headers: List[str] = [
+            'Op type',
+            'Op',
+            'Param size',
+            'FLOPs',
+        ]
+
+        return tabulate(node_summaries, headers=headers, stralign='right')
+
+
+def counter_pass(module: torch.fx.GraphModule, *args, verbose=False) -> torch.fx.GraphModule:
+    """A pass that counts the number of FLOPs and parameters in a model.
+
+    Parameters
+    ----------
+    module: torch.fx.GraphModule
+        The module to be profiled.
+
+    verbose: bool
+        Whether to print the summary of the profiled statistics. Default: False.
+
+    Returns
+    -------
+    module: torch.fx.GraphModule
+        The same module as the input.
```

Comment:
There could be another function to calculate these metrics.

Reply:
I think returning a dict might be confusing because it's not clear to which granularity should we make these results.

May it be fx.node or nn.Module. should depend on the clear use case though.

Reply:
> I think returning a dict might be confusing because it's not clear to which granularity should we make these results.
> 
> May it be fx.node or nn.Module. should depend on the clear use case though.

Good idea! `nn.Module` or op level is better!

Reply:
Hopefully now it returns what is correct.